src/main/java/com/google/netpcapanalysis/dao/MaliciousIPDaoImpl.java

```diff
@@ -73,4 +78,65 @@ else if (result.getBody().contains(REQUEST_LIMIT))
         return data;
     }
     
+    private String isMalicious(String ip)
+    {
+        HttpResponse<String> apiResult;
+        String result;
+
+        String searchDB = ipCache.searchMaliciousDB(ip);
+        if(searchDB.equalsIgnoreCase(Flagged.TRUE))
+        {
+            result = Flagged.TRUE;
+        }
+        else if(searchDB.equalsIgnoreCase(Flagged.FALSE)){
+            result = Flagged.FALSE;
+        }
+        else{
+            System.out.println("Else statement");
+            try {
+                apiResult = Unirest.get("https://signals.api.auth0.com/badip/" + ip)
+                        .header("X-Auth-Token", AUTH0_API_KEY)
+                        .asString();
+           
+                if(apiResult.getBody().equalsIgnoreCase(FLAGGED_FALSE))
+                {
+                    result = Flagged.FALSE;
+                    MaliciousPacket tempPacket = new MaliciousPacket(ip, result);
+                    ipCache.setMaliciousIPObjects(tempPacket);
+                }
+                else if(apiResult.getBody().equalsIgnoreCase(FLAGGED_TRUE))
+                {
+                    result = Flagged.TRUE; 
+                    MaliciousPacket tempPacket = new MaliciousPacket(ip, result);
+                    ipCache.setMaliciousIPObjects(tempPacket);
+                }
+                else if (apiResult.getBody().contains(REQUEST_LIMIT))
+                {
+                    result = Flagged.UNKNOWN; 
+                }
+                else{
+                    result = Flagged.ERROR; 
+                }
+    
+            } catch (UnirestException e) {
+                System.out.println("IP lookup failed");
+                result = Flagged.ERROR;
+                // TODO Auto-generated catch block
+                e.printStackTrace();
+            }
+        }
+        return result;
+    }
+    
+    public void run(String searchEntity){
+        PCAPDao datastore = new PCAPDaoImpl();
+        ArrayList<PCAPdata> allData = datastore.getPCAPObjects(searchEntity); 
+        ArrayList<PCAPdata> uniqueData = datastore.getUniqueIPs(allData); 
+        String flaggeValue;
+
+        for(PCAPdata packet : uniqueData){
+            flaggeValue = isMalicious(packet.destination);
+            datastore.updateFlagged(searchEntity, packet, flaggeValue);
```

Comment:
You should decouple the flagged information from the data as well. I believe your data store could potentially have duplicate IPs too. Either way this information maps a IP -> Flagged and should not be coupled with the original PCAPdata. If you want to store it it permanently other than Cache, then it should be in a separate entity, where the advantage is if you have a new PCAP file with same IPs it would be the same source of truth and wont go stale.

Reply:
Will this be the same for location and domain lookup? In that case should remove (Domain, Location, Freq, Flagged) fields from datastore since.

Reply:
Yes it applies to all of them. Think of it this way, the stuff that is in the PCAP is static portion of your data, whereas the other things you compute on the fly are dynamic (i.e. they could potentially change in future). Also it is all based on Dest IP right, and in real data you may not have unique Dest IPs e.g. you could be talking to the same Dest IP with different protocols. I would say by removing this it would actually make your modules lot simpler and extensible, and you can use the cache better. But if it is too much work for you folks then we can keep it as it is.